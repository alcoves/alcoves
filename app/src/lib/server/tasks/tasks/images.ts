import { mkdtemp, rm } from "node:fs/promises";
import { tmpdir } from "os";
import { join } from "path";
import { Worker } from "bullmq";
import sharp from "sharp";
import { v4 as uuid } from "uuid";
import { db } from "../../db/db";
import { env } from "$lib/server/utilities/env";
import { assetThumbnails } from "../../db/schema";;
import { ImageTasks, bullConnection, imageProcessingQueue } from "../queues";
import { downloadObject, uploadFileToS3 } from "$lib/server/utilities/s3";

export interface ImageProxyJobData {
	assetId: string;
	sourceKey: string;
	sourceBucket: string;
}

interface ImageProxyJob {
	name: ImageTasks;
	data: ImageProxyJobData;
}

async function main() {
	const worker = new Worker(
		imageProcessingQueue.name,
		async (job: ImageProxyJob) => {
			const tmpDir = await mkdtemp(join(tmpdir(), "alcoves-"));
			console.info(`Created temporary directory: ${tmpDir}`);

			try {
				if (job.name === ImageTasks.GENERATE_IMAGE_PROXIES) {
					const sourceImage = await downloadObject({
						localDir: tmpDir,
						key: job.data.sourceKey,
						bucket: job.data.sourceBucket,
					});

					const filepath = join(tmpDir, "image.avif");

					const sizes = [
						{ size: "sm", width: 400, quality: 50 },
						{ size: "md", width: 800, quality: 65 },
						{ size: "lg", width: 1200, quality: 75 },
					];

					for (const size of sizes) {
						const compressedImage = await sharp(sourceImage)
							.resize({ width: 400 })
							.avif({ quality: 65 })
							.rotate() // auto-rotate based on EXIF data
							.toFile(filepath);

						const proxyStorageId = uuid();
						const storageBucket = env.ALCOVES_OBJECT_STORE_DEFAULT_BUCKET;
						const storageKey = `DEBUG/${proxyStorageId}`;

						const uploadedObject = await uploadFileToS3({
							filePath: filepath,
							key: storageKey,
							bucket: storageBucket,
							contentType: "image/avif",
						});

						const assetImageProxy = await db
							.insert(assetThumbnails)
							.values({
								assetId: job.data.assetId,
								size: compressedImage.size,
								width: compressedImage.width,
								height: compressedImage.height,
								storageBucket: storageBucket,
								storageKey: storageKey,
							})
							.returning();
					}

					const compressedImage = await sharp(sourceImage)
						.resize({ width: 400 })
						.avif({ quality: 65 })
						.rotate() // auto-rotate based on EXIF data
						.toFile(filepath);

					const proxyStorageId = uuid();
					const storageBucket = env.ALCOVES_OBJECT_STORE_DEFAULT_BUCKET;
					const storageKey = `DEBUG/${proxyStorageId}`;

					const uploadedObject = await uploadFileToS3({
						filePath: filepath,
						key: storageKey,
						bucket: storageBucket,
						contentType: "image/avif",
					});

					const assetThumbnail = await db
						.insert(assetThumbnails)
						.values({
							assetId: job.data.assetId,
							size: compressedImage.size,
							width: compressedImage.width,
							height: compressedImage.height,
							storageBucket: storageBucket,
							storageKey: storageKey,
						})
						.returning();
				}

				// switch (job.name) {
				// 	case ImageTasks.GENERATE_IMAGE_PROXIES:
				// 		console.log("Image proxy Job");
				// 		break;
				// 	case ImageTasks.FETCH_IMAGE_METADATA:
				// 		console.info("Downloading image", job.data);
				// 		const imagePath = await downloadObject({
				// 			localDir: tmpDir,
				// 			key: job.data.key,
				// 			bucket: job.data.bucket,
				// 		});
				// 		console.info("Parsing metadata");
				// 		const metadata = await sharp(imagePath).metadata();
				// 		// console.log(JSON.stringify(metadata, null, 2))

				// 		const compressedImage = await sharp(imagePath)
				// 			.resize({
				// 				width: 400,
				// 			})
				// 			.avif({
				// 				quality: 50,
				// 				// progressive: true,
				// 			})
				// 			.toBuffer();

				// 		console.info("Uploading compressed image");
				// 		const url = await uploadBufferToS3({
				// 			imageBuffer: compressedImage,
				// 			bucket: job.data.bucket,
				// 			key: "test.jpg",
				// 		});

				// 		// Update database
				// 		break;
				// 	default:
				// 		break;
				// }
			} catch (error) {
				console.error("Error processing image", error);
				throw error;
			} finally {
				console.info("Removing temporary directory");
				rm(tmpDir, { recursive: true, force: true });
			}
		},
		{
			connection: bullConnection,
			concurrency: Number.parseInt(env.ALCOVES_TASK_WORKER_CONCURRENCY),
		},
	);

	worker.on("completed", (job) => {
		console.log(`${job.id} has completed!`);
	});

	worker.on("failed", (job, err) => {
		console.log(`${job?.id} has failed with ${err.message}`);
	});

	console.info(
		`Starting worker: ${worker.name} for queue: ${imageProcessingQueue.name}`,
	);
}

export default main;
