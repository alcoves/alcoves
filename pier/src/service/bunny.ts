import axios from 'axios'

export function purgeURL(url: string) {
  const accessKey = process.env.BUNNY_ACCESS_KEY
  if (!accessKey) return console.error('invalid bunny access key')

  console.log(`Purging URL ${url} from cdn`)
  return axios({
    method: 'POST',
    headers: { AccessKey: accessKey },
    url: `https://api.bunny.net/purge?url=${url}`,
  })
}
