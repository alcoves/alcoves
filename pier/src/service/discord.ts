import axios from 'axios'

const webhookURL = process.env.DISCORD_WEBHOOK

export async function discordWebHook(message: string) {
  if (!webhookURL) return console.error('Discord webhook url not specified')
  if (process.env.NODE_ENV === 'production') {
    return axios.post(webhookURL, {
      content: message,
    })
  }
}
