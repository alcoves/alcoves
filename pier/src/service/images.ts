import sharp from 'sharp'
import mime from 'mime-types'
import s3 from '../config/s3'
import { ManagedUpload } from 'aws-sdk/clients/s3'

interface DataURIScheme {
  data: string
  encoding: string
  contentType: string
}

// Parses https://en.wikipedia.org/wiki/Data_URI_scheme
export function parseDataURIScheme(dataURIScheme: string): DataURIScheme | null {
  const withoutData = dataURIScheme.split('data:')[1] || ''
  const [contentType, encoding, data] = withoutData.split(/;|,/)

  if (!data) return null
  if (encoding !== 'base64') return null

  return {
    contentType,
    encoding,
    data,
  }
}

export async function optimizeUserAvatar(dataUri: DataURIScheme) {
  try {
    const imageBuffer = Buffer.from(dataUri.data, 'base64')
    const image = await sharp(imageBuffer)
      .resize(250, 250, {
        fit: 'contain',
      })
      .jpeg({
        quality: 80,
        progressive: true,
      })
      .toBuffer()
    return image
  } catch (error) {
    console.error(error)
    throw new Error('Failed to optimize user avatar')
  }
}
