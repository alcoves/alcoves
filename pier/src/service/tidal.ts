import axios from 'axios'
import { Video } from '@prisma/client'
import { archiveBucket, cdnBucket } from '../config/s3'

export function dispatchJob(route: string, data: any) {
  const dispatchURL = new URL(route, process.env.TIDAL_URL).toString()
  console.log('dispatchURL', dispatchURL)

  const apiKey = process.env.TIDAL_API_KEY
  if (!apiKey) throw new Error('Invalid TIDAL_API_KEY')
  return axios.post(dispatchURL, data, { headers: { 'x-api-key': apiKey } })
}

export function parseDimensions(metadata): { width: number; height: number } {
  let width: number = metadata?.video[0]?.width || 0
  let height: number = metadata?.video[0]?.height || 0

  if (metadata?.video[0]?.tags?.rotate) {
    const rotateInt = parseInt(metadata?.video[0]?.tags?.rotate)
    switch (rotateInt) {
      case 90:
        width = metadata?.video[0]?.height
        height = metadata?.video[0]?.width
        break
      case -90:
        width = metadata?.video[0]?.height
        height = metadata?.video[0]?.width
    }
  }

  return { width, height }
}

export async function tidalVideoCreate(video: Video) {
  try {
    const sourceInput = `s3://${archiveBucket}/${video.archivePath}`

    await dispatchJob('/videos/metadata', {
      assetId: video.id,
      input: sourceInput,
    })

    await dispatchJob('/videos/thumbnails', {
      width: 854,
      height: 480,
      fit: 'cover',
      assetId: video.id,
      input: sourceInput,
      output: `s3://${cdnBucket}/v/${video.id}/thumbnail.webp`,
    })

    await dispatchJob('/videos/transcodes/adaptive', {
      assetId: video.id,
      input: sourceInput,
      output: `s3://${cdnBucket}/v/${video.id}`,
    })
  } catch (error) {
    console.error(error)
    throw error
  }
}
