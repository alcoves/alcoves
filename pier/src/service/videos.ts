export function parseFramerate(r_frame_rate: string): number {
  let framerate: number

  if (r_frame_rate.includes('/')) {
    // Probably like 60/1 or something
    const [frames, time] = r_frame_rate.split('/')
    framerate = parseFloat(frames) / parseFloat(time)
  } else {
    // Probably like 23.976
    framerate = parseFloat(r_frame_rate)
  }

  if (framerate <= 15.1 && framerate >= 14.9) {
    return 15
  }

  if (framerate <= 25.1 && framerate >= 23.9) {
    return 24
  }

  if (framerate <= 30.1 && framerate >= 29.9) {
    return 30
  }

  if (framerate <= 60.1 && framerate >= 59.9) {
    return 60
  }

  if (framerate > 60.1) {
    return 60
  }

  return Math.round(framerate)
}
