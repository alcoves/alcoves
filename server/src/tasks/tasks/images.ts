import sharp from "sharp";
import { join } from "path";
import { tmpdir } from "os";
import { Worker } from "bullmq";
import { env } from "../../lib/env";
import { rm, mkdtemp } from "node:fs/promises";
import { downloadObject, uploadBufferToS3 } from "../../lib/s3";
import { ImageTasks, bullConnection, imageProcessingQueue } from "../queues";

async function main() {
	const worker = new Worker(
		imageProcessingQueue.name,
		async (job) => {
			const tmpDir = await mkdtemp(join(tmpdir(), "alcoves-"));
			console.info(`Created temporary directory: ${tmpDir}`);

			try {
				switch (job.name) {
					case ImageTasks.GENERATE_IMAGE_PROXIES:
						console.log("Image proxy Job");
						break;
					case ImageTasks.FETCH_IMAGE_METADATA:
						console.info("Downloading image", job.data);
						const imagePath = await downloadObject({
							localDir: tmpDir,
							key: job.data.key,
							bucket: job.data.bucket,
						});
						console.info("Parsing metadata");
						const metadata = await sharp(imagePath).metadata();
						// console.log(JSON.stringify(metadata, null, 2))

						const compressedImage = await sharp(imagePath)
							.resize({
								width: 400,
							})
							.avif({
								quality: 50,
								// progressive: true,
							})
							.toBuffer();

						console.info("Uploading compressed image");
						const url = await uploadBufferToS3({
							imageBuffer: compressedImage,
							bucket: job.data.bucket,
							key: "test.jpg",
						});

						// Update database
						break;
					default:
						break;
				}
			} catch (error) {
				console.error("Error processing image", error);
				throw error;
			} finally {
				console.info("Removing temporary directory");
				rm(tmpDir, { recursive: true, force: true });
			}
		},
		{
			connection: bullConnection,
			concurrency: Number.parseInt(env.ALCOVES_TASK_WORKER_CONCURRENCY || "1"),
		},
	);

	worker.on("completed", (job) => {
		console.log(`${job.id} has completed!`);
	});

	worker.on("failed", (job, err) => {
		console.log(`${job?.id} has failed with ${err.message}`);
	});

	console.info(`Starting worker: ${worker.name} for queue: ${imageProcessingQueue.name}`);
}

export default main;
