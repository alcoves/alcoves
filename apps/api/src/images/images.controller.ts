import { Response } from 'express'
import {
  Controller,
  Get,
  Post,
  Body,
  Patch,
  Param,
  Delete,
  Res,
  Query,
} from '@nestjs/common'
import { ImagesService } from './images.service'
import { CreateImageDto } from './dto/create-image.dto'
import { UpdateImageDto } from './dto/update-image.dto'
import { GetImageParamsDto, GetImageQueryDto } from './dto/get-image-dto'

@Controller('images')
export class ImagesController {
  constructor(private readonly imagesService: ImagesService) {}

  @Post()
  create(@Body() createImageDto: CreateImageDto) {
    return this.imagesService.create(createImageDto)
  }

  @Get()
  findAll() {
    return this.imagesService.findAll()
  }

  @Get(':id.:format')
  findOne(
    @Param() params: GetImageParamsDto,
    @Query() query: GetImageQueryDto,
    @Res() res: Response
  ) {
    return this.imagesService.findOne(params, query, res)
  }

  @Patch(':id')
  update(@Param('id') id: string, @Body() updateImageDto: UpdateImageDto) {
    return this.imagesService.update(+id, updateImageDto)
  }

  @Delete(':id')
  remove(@Param('id') id: string) {
    return this.imagesService.remove(+id)
  }
}
