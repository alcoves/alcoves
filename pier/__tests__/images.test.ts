import sharp from 'sharp'
import { optimizeUserAvatar, parseDataURIScheme } from '../src/service/images'

describe('parseDataURIScheme', () => {
  const successCases = [
    {
      recieved: 'data:image/jpg;base64,123',
      expected: {
        data: '123',
        encoding: 'base64',
        contentType: 'image/jpg',
      },
    },
    {
      recieved: 'data:image/png;base64,1231251251241241251231',
      expected: {
        data: '1231251251241241251231',
        encoding: 'base64',
        contentType: 'image/png',
      },
    },
  ]

  test.each(successCases)('successfully parses %i', ({ recieved, expected }) => {
    expect(parseDataURIScheme(recieved)).toMatchObject(expected)
  })

  const failureCases = [['data2:image/jpg;base64,123'], ['data:image/jpg;notbase64,123']]
  test.each(failureCases)('fails to parse %i', recieved => {
    expect(parseDataURIScheme(recieved)).toBe(null)
  })
})

describe('optimizeUserAvatar', () => {
  test('', async () => {
    const dataURI =
      'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAABpklEQVR4AcXVA48dYRTH4be2HdZ2o9qIUZvfow5qR3UbVev9Ftc3zjpYG/Xpr+YZvp17kmc8c/5j47lOyQgcxj2k0AqBoB1pPMUJTDLW6pRMx330QDz6gFKsDtN4IC7iPSSEPIz323wKshBL6rDCa/MFaIBY1o1tbs1n2Gqu6MIyrfkgZCAuynEFe7ASa7ED55GAuKjGiH8FuAyBpgo70dvlKq5GHALN/T93mo13EEUxhvl4iPvhDkTxEYt+3eE+RFGK/iZAuYR4/n2j0XijXXaMCPEt6afdDrzDRMPgAESxx8KXdDVEccIweKisrEEfY6P0t+OpYZBSVt60+D85r/RIGwZtyspDFgPsUHq0Gwai2GQxwFq1TyTFFzPXAfYoAVqjCnBFCZCKonlvlCsB7vk5kPyT+347IYrD/zcAPzBUKfv2YMQ/D24lAD8wFHv5JdsPwJmhFKJ4j+n2A/DvwB5UQTS4aCgrAbAJh3ATNRAXWQz0/cD52lbXgCmGykWABiwwVC4CZDDdUFEHeIfLGGSoKAO8wX3MjuJbL2hDCg9xAKONx/oE7ZLWqbmf4UoAAAAASUVORK5CYII='
    const parsedDataURIScheme = parseDataURIScheme(dataURI)
    if (parsedDataURIScheme) {
      const imageBuffer = await optimizeUserAvatar(parsedDataURIScheme)
      const metadata = await sharp(imageBuffer).metadata()
      expect(metadata).toMatchSnapshot()
    }
  })
})
