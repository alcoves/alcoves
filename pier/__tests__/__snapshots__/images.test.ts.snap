// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`optimizeUserAvatar  1`] = `
Object {
  "channels": 3,
  "chromaSubsampling": "4:2:0",
  "density": 72,
  "depth": "uchar",
  "format": "jpeg",
  "hasAlpha": false,
  "hasProfile": false,
  "height": 250,
  "isProgressive": true,
  "size": 5052,
  "space": "srgb",
  "width": 250,
}
`;
